cloud/blockstore/libs/storage/disk_agent/rdma_target.cpp

```diff
@@ -575,11 +578,21 @@ class TRequestHandler final
                     << error.GetMessage() << " (" << error.GetCode() << ")");
         }
 
-        TStackVec<IOutputStream::TPart> parts;
-        parts.reserve(blocks.BuffersSize());
+        bool allZeroes = false;
+        if (requestDetails.SkipVoidBlocks) {
+            allZeroes = AllOf(
+                blocks.GetBuffers(),
+                [](const TString& buffer)
+                { return IsAllZeroes(buffer.data(), buffer.size()); });
+        }
+        proto.SetAllZeroes(allZeroes);
 
-        for (const auto& buffer: blocks.GetBuffers()) {
-            parts.emplace_back(TStringBuf(buffer));
+        TStackVec<IOutputStream::TPart> parts;
+        if (!allZeroes) {
```

Comment:
With this PR the whole data buffer including zeroes will still be written using RDMA to the client.
The actual send of RDMA buffer back to client is done [here ](https://github.com/ydb-platform/nbs/blob/458fa10fd187afe0598dd5ed9824bccc460aef8b/cloud/blockstore/libs/rdma/impl/server.cpp#L893)
And the buffer includes  `header + proto + data`. 

The rdma lib itself is actually not aware of what is encoded in the payload. If for example the user wants to read 512 bytes. The rdma lib will allocate 4k aligned size of buffer which is enough to contain `header+proto+512 bytes` and it will RDMA it back regardless of the content of the buffer. Even if there is an error  encoded in the buffer and no data needs to be written back it will still RDMA back the whole buffer. This is not very efficient but it's simpler to implement.

So to save bytes on the line you actually need to change the rdma lib.

Right now I am working on [zero-buffer-copy](https://github.com/ydb-platform/nbs/commit/6cf475dedce82dbbd0a3c301e9411ac32202306d) optimization which we need on agent since there are plenty buffer copies which become bottleneck in setups with 100Gbit nics doing sequential read/write with large blocks. I still need to test it and finish the implementation and then I will open PR. Would be great if you can wait for zero copy and then merge void buffers transfers  to avoid conflicts and repeating the RDMA tests

Reply:
Все правильно, по сети передается header + proto + data.
Количество байтов, которые нужно передать по сети формирует NRdma::TProtoMessageSerializer::Serialize()
А мы в свою очередь при сериализации  отдаем пустой список блоков, и размер data оказывается равным нулю.

Reply:
Обсудили в личке. Да, циферка, что возвращает Serialize() не влияет. Буду упрощать ПР и детектировать нули на стороне клиента